bigquery_etl/routine/parse_routine.py

```diff
@@ -327,28 +299,52 @@ def sub_local_routines(test, project, raw_routines=None, stored_procedure_test=F
     if raw_routines is None:
         raw_routines = read_routine_dir()
 
-    sql = prepend_routine_usage_definitions(test, project, raw_routines)
-
-    for name, routine in raw_routines.items():
-        if name in sql:
-            for defn in routine.definitions:
-                match = PERSISTENT_UDF_RE.match(defn)
-                dataset, name = match.group(1), match.group(2)
-                replace_name = f"{dataset}_{name}"
-                if stored_procedure_test:
-                    replace_name = f"{GENERIC_DATASET}.{replace_name}"
-                sql = re.sub(
-                    rf"`?(?:`?{routine.project}`?\.)?{dataset}.{name}`?",
-                    replace_name,
-                    sql,
-                )
+    routine_ids = []
+    for routine_id in routine_usages_in_text(test, project_dir):
+        routine_ids = accumulate_dependencies(routine_ids, raw_routines, routine_id)
+        # Add a routine after its dependencies so the routines get declared in the correct order.
+        if routine_id not in routine_ids:
+            routine_ids.append(routine_id)
+
+    routines = [
+        raw_routines[routine_id]
+        for routine_id in routine_ids
+        if routine_id in raw_routines
+    ]
+
+    sql = test
+    if routines:
+        routines_sql = "\n\n".join(
+            definition
+            for routine in routines
+            for definition in routine.definitions
+            if definition not in sql
+        )
+        sql = routines_sql + "\n\n" + sql
+
+    for routine in routines:
+        for definition in routine.definitions:
+            routine_match = PERSISTENT_UDF_RE.match(definition)
+            routine_project = routine_match.group(1) or routine.project
+            routine_dataset = routine_match.group(2)
+            routine_name = routine_match.group(3)
+            replace_routine_name = re.sub(
+                r"\W", "_", f"{routine_project}_{routine_dataset}_{routine_name}"
+            )
```

Comment:
I was hoping this regex logic could be avoided and simply use `replace_routine_name = re.sub(r"\W", "_", routine.id)`, but it turns out there's one UDF file that doesn't follow the one-UDF-per-file convention and defines multiple UDFs: [sql/moz-fx-data-shared-prod/udf/safe_crc32_uuid/udf.sql](https://github.com/mozilla/bigquery-etl/blob/main/sql/moz-fx-data-shared-prod/udf/safe_crc32_uuid/udf.sql)